app/services/dqt/retrieve_training_instance.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Dqt
+  class RetrieveTrainingInstance
+    include ServicePattern
+
+    def initialize(trainee:)
+      @trainee = trainee
+    end
+
+    def call
+      raise(DqtNoTrainingInstanceError) if training_instances.blank? || training_instance.nil?
+
+      training_instance
+    end
+
+    attr_reader :trainee
+
+  private
+
+    def training_instance
+      @training_instance ||= training_instances.find do |training|
+        training.dig("provider", "ukprn") == trainee.provider.ukprn &&
+          training["programmeType"] == programme_type(trainee)
+      end
+    end
+
+    def programme_type(trainee)
+      Dqt::Params::TrnRequest.new(trainee:).params.dig("initialTeacherTraining", "programmeType")
```

Comment:
I remember suggesting `Dqt::Params::TrnRequest.new(trainee:).params` to get the `programmeType` but since you're only using one value I'm thinking `Dqt::Params::TrnRequest::PROGRAMME_TYPE[trainee.training_route]` might be bit nicer. I don't mind either way.

Reply:
your's is defs nicer, updated!